cli/src/host_perf_check.rs

```diff
@@ -0,0 +1,95 @@
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::error::PerfCheckError;
+use log::info;
+use nix::unistd;
+use polkadot_node_core_pvf::sp_maybe_compressed_blob;
+use sc_service::BasePath;
+use service::kusama_runtime;
+use std::{
+	fs::{self, OpenOptions},
+	io::{self, Read, Write},
+	path::Path,
+	time::{Duration, Instant},
+};
+
+fn is_perf_check_done(path: &Path) -> io::Result<bool> {
+	let host_name_max_len = unistd::SysconfVar::HOST_NAME_MAX as usize;
+
+	let mut host_name = vec![0u8; host_name_max_len];
+	let mut buf = host_name.clone();
+	unistd::gethostname(&mut host_name).map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
+
+	let file = match fs::File::open(path) {
+		Ok(file) => file,
+		Err(err) if err.kind() == io::ErrorKind::NotFound => return Ok(false),
+		Err(err) => return Err(err),
+	};
+	let mut reader = io::BufReader::new(file);
+
+	reader.read_exact(&mut buf)?;
+
+	Ok(host_name == buf)
+}
+
+fn save_check_passed_file(path: &Path) -> io::Result<()> {
+	let host_name_max_len = unistd::SysconfVar::HOST_NAME_MAX as usize;
+	let mut host_name = vec![0u8; host_name_max_len];
+	unistd::gethostname(&mut host_name).map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
+
+	let mut file = OpenOptions::new().truncate(true).create(true).write(true).open(path)?;
+
+	file.write(&host_name)?;
+
+	Ok(())
+}
+
+pub fn host_perf_check(result_cache_path: Option<BasePath>) -> Result<(), PerfCheckError> {
+	const PERF_CHECK_TIME_LIMIT: Duration = Duration::from_secs(20);
+	const CODE_SIZE_LIMIT: usize = 1024usize.pow(3);
+	const CHECK_PASSED_FILE_NAME: &str = ".perf_check_passed";
+	let wasm_code = kusama_runtime::WASM_BINARY.ok_or(PerfCheckError::WasmBinaryMissing)?;
+
+	let check_passed_file_path =
+		result_cache_path.map(|path| path.path().join(CHECK_PASSED_FILE_NAME));
+
+	if let Some(ref path) = check_passed_file_path {
+		if let Ok(true) = is_perf_check_done(path) {
+			info!("Performance check skipped: already passed");
+			return Ok(())
+		}
+	}
+
+	info!("Running the performance check...");
+	let start = Instant::now();
+
+	// Recreate the pipeline from the pvf prepare worker.
+	let code = sp_maybe_compressed_blob::decompress(wasm_code, CODE_SIZE_LIMIT)
+		.or(Err(PerfCheckError::CodeDecompressionFailed))?;
+	let blob = polkadot_node_core_pvf::prevalidate(code.as_ref()).map_err(PerfCheckError::from)?;
+	let _ = polkadot_node_core_pvf::prepare(blob).map_err(PerfCheckError::from)?;
+
+	let elapsed = start.elapsed();
+	if elapsed <= PERF_CHECK_TIME_LIMIT {
+		info!("Performance check passed, elapsed: {:?}", start.elapsed());
```

Comment:
This deviates a bit from the intended output of `good`, `ok`, `fail`, so I think this should be a `println!` in any case and do some human representation of the outcome, maybe some emojis.

Reply:
Why `println`?

Reply:
Because the whole purpose of this command is to show output to the user.

Reply:
It's explicitly set to `info` in `LoggerBuilder`, I don't think there's a possibility of missing it.

Reply:
The 3 level case result metric, `good`, `bad`, `fail` is still pending impl.

Reply:
We use `info` almost everywhere for this kind of output. However, if we for example want to print stuff that should be redirected, we use `println!`. The nice thing is that we log to stderr and println goes to stdout. So you can redirect to a file without getting any informational output in between.

TDLR: info is fine.